app/domain/authentication/authn_gcp/validate_resource_restrictions_match_jwt.rb

```diff
@@ -0,0 +1,60 @@
+module Authentication
+  module AuthnGcp
+
+    # This class is responsible of validating resource restrictions values against JWT token.
+    # the assumption is that resource restrictions list contains only permitted types which already validated
+    # in previous steps
+    ValidateResourceRestrictionsMatchJWT = CommandClass.new(
+      dependencies: {
+        logger: Rails.logger
+      },
+      inputs:       %i(resource_restrictions decoded_token restriction_prefix)
+    ) do
+
+      def call
+        @logger.debug(LogMessages::Authentication::AuthnGcp::ValidatingResourceRestrictionsValues.new)
+        validate
+        @logger.debug(LogMessages::Authentication::AuthnGcp::ValidatedResourceRestrictionsValues.new)
+      end
+
+      private
+
+      def validate
+        @resource_restrictions.each do |r|
+          resource_type = r.type
+          resource_value = r.value
+
+          case remove_prefix(resource_type)
+          when PROJECT_ID_RESTRICTION_NAME
+            unless resource_value == @decoded_token.project_id
+              raise Errors::Authentication::AuthnGcp::InvalidResourceRestrictions.new(resource_type)
+            end
+            @logger.debug(LogMessages::Authentication::ValidatedResourceRestrictionValue.new(resource_type))
+
+          when INSTANCE_NAME_RESTRICTION_NAME
+            unless resource_value == @decoded_token.instance_name
+              raise Errors::Authentication::AuthnGcp::InvalidResourceRestrictions.new(resource_type)
+            end
+            @logger.debug(LogMessages::Authentication::ValidatedResourceRestrictionValue.new(resource_type))
+
+          when SERVICE_ACCOUNT_ID_RESTRICTION_NAME
+            unless resource_value == @decoded_token.service_account_id
+              raise Errors::Authentication::AuthnGcp::InvalidResourceRestrictions.new(resource_type)
+            end
+            @logger.debug(LogMessages::Authentication::ValidatedResourceRestrictionValue.new(resource_type))
+
+          when SERVICE_ACCOUNT_EMAIL_RESTRICTION_NAME
+            unless resource_value == @decoded_token.service_account_email
+              raise Errors::Authentication::AuthnGcp::InvalidResourceRestrictions.new(resource_type)
+            end
+            @logger.debug(LogMessages::Authentication::ValidatedResourceRestrictionValue.new(resource_type))
+          end
```

Comment:
i'd like to suggest another solution, similar to what we have in `authn-azure`. Instead of having a switch case that runs over all the options and performs a similar block of code, we can add another method called `jwt_token_resources` that will create a list of `ResourceRestrictions` objects. Then, the `validate` method will verify that it contains all of the objects of `resource_restrictions`.

in `authn-azure` it looks like this:
```
def validate_resource_restrictions_match_request
        resource_restrictions.resources.each do |resource_from_role|
          resource_from_request = resources_from_request.find { |resource| resource == resource_from_role }
          unless resource_from_request
            raise Errors::Authentication::AuthnAzure::InvalidResourceRestrictions, resource_from_role.type
          end
        end
        @logger.debug(LogMessages::Authentication::ValidatedResourceRestrictions.new)
      end

      # xms_mirid is a term in Azure to define a claim that describes the resource
      # that holds the encoding of the instance's among other details the subscription_id,
      # resource group, and provider identity needed for authorization.
      # xms_mirid is one of the fields in the JWT token. This function will extract the relevant information from
      # xms_mirid claim and populate a representative hash with the appropriate fields.
      def resources_from_request
        return @resources_from_request if @resources_from_request

        @resources_from_request = [
          AzureResource.new(
            type: "subscription-id",
            value: xms_mirid.subscriptions
          ),
          AzureResource.new(
            type: "resource-group",
            value: xms_mirid.resource_groups
          )
        ]

        # determine which identity is provided in the token. If the token is
        # issued to a user-assigned identity then we take the identity name.
        # If the token is issued to a system-assigned identity then we take the
        # Object ID of the token.
        if xms_mirid.providers.include? "Microsoft.ManagedIdentity"
          @resources_from_request.push(
            AzureResource.new(
              type: "user-assigned-identity",
              value: xms_mirid.providers.last
            )
          )
        else
          @resources_from_request.push(
            AzureResource.new(
              type: "system-assigned-identity",
              value: @oid_token_field
            )
          )
        end
        @logger.debug(LogMessages::Authentication::AuthnAzure::ExtractedResourceRestrictionsFromToken.new)
        @resources_from_request
      end
```

One issue that rises from this is that having a `ResourceRestriction` objects list of the JWT is quite weird because the JWT doesn't hold restrictions. it holds resources. To that i would say that the `ResourceRestriction` object does not really represent a resource restriction, rather than a resource (we restrict on it but the object itself reperesents a resource). So i would change the code to the above and refactor `ResourceRestriction` to `Resource`.

wdyt?

Reply:
Thanks 🙏  for the elaborated suggestion, i liked your approach but it still will have a duplicate code block, for example in your solution the duplication will be in the Array creation:
```
      def resources_from_request
        return @resources_from_request if @resources_from_request

        @resources_from_request = [
          ResourceRestriction.new(
            type: PROJECT_ID_RESTRICTION_NAME,
            value: @decoded_token.project_id
          ),
          ResourceRestriction.new(
            type: INSTANCE_NAME_RESTRICTION_NAME,
            value: @decoded_token.instance_name
          ),
          ResourceRestriction.new(
            type: SERVICE_ACCOUNT_ID_RESTRICTION_NAME,
            value: @decoded_token.service_account_id
          ),
          ResourceRestriction.new(
            type: SERVICE_ACCOUNT_EMAIL_RESTRICTION_NAME,
            value: @decoded_token.service_account_email
          )
        ]
      end
```

What do you think about the following alternative:
```
      def validate
        @resource_restrictions.each do |r|
          restriction_type = r.type
          restriction_value = r.value
          restriction_type_without_prefix = remove_prefix(restriction_type)
          restriction_value_from_token = get_value_from_token(restriction_type_without_prefix)
          unless restriction_value == restriction_value_from_token
            raise Errors::Authentication::AuthnGcp::InvalidResourceRestrictions.new(restriction_type)
          end
          @logger.debug(LogMessages::Authentication::ValidatedResourceRestrictionValue.new(restriction_type))
        end
      end

      def remove_prefix(restriction_type)
        restriction_type_without_prefix = restriction_type.sub(@restriction_prefix, "")
        restriction_type_without_prefix
      end

      def get_value_from_token(restriction_type_without_prefix)
        restriction_value_from_token = nil

        case restriction_type_without_prefix
        when PROJECT_ID_RESTRICTION_NAME
          restriction_value_from_token = @decoded_token.project_id
        when INSTANCE_NAME_RESTRICTION_NAME
          restriction_value_from_token = @decoded_token.instance_name
        when SERVICE_ACCOUNT_ID_RESTRICTION_NAME
          restriction_value_from_token = @decoded_token.service_account_id
        when SERVICE_ACCOUNT_EMAIL_RESTRICTION_NAME
          restriction_value_from_token = @decoded_token.service_account_email
        end

        restriction_value_from_token
      end
```

I think it is more readable, and the code duplication here is minimal.

WDYT?

Reply:
yes, i'm ok with your suggestion. if you do go with it then you don't need the `==` method implementation in `ResourceRestriction` class (i added that to compare between `ResourceRestriction` objects but you don't do it here).

also please change `remove_prefix` to `restriction_type_without_prefix` or `non_prefixed_restriction_type`